### PR TITLE
DR-1581: Add versioning to newly created buckets

### PR DESCRIPTION
**Pull request by fboulnois**
This addition is to comply with FISMA / FEDRAMP policies.

### CODE REVIEW

File: src/main/java/bio/terra/service/resourcemanagement/google/GoogleBucketService.java [hunk 1]
```diff
@@ -35,6 +36,8 @@
     private final GcsConfiguration gcsConfiguration;
     private final ConfigurationService configService;
 
+    private static final Long RETENTION_PERIOD = Duration.ofDays(30).getSeconds();
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/resourcemanagement/google/GoogleBucketService.java]: unnecessary use of boxed type
```suggestion
    private static final long RETENTION_PERIOD = Duration.ofDays(30).getSeconds();
```

**Review by pshapiro4broad — APPROVED** [review]
This looks OK to me. What will we do if someone wants to delete a bucket (or an object in a bucket) whose data is protected under this policy, will we manually remove the policy at that point?

File: src/main/java/bio/terra/service/resourcemanagement/google/GoogleBucketService.java (hunk 1 above)
Comment (fboulnois) [reply src/main/java/bio/terra/service/resourcemanagement/google/GoogleBucketService.java]: I like the idea of having to manually opt-out to remove the policy -- it makes everyone aware of the danger of doing that. Also, good catch on the boxed type.

**Comment by ddietterich** [comment]
This is not what I expected as the solution. It is probably fine in the short term, but we do need to be able to do deletes and this will prevent that.

I think we we want is to enable object versioning and use a lifecycle rule to delete non-current versions that are older than 30 days. If you delete a file, it does not go away - it becomes non-current until 30 days have gone by.

See https://cloud.google.com/storage/docs/managing-lifecycles#deletion-example

**Comment by pshapiro4broad** [comment]
> I think we we want is to enable object versioning and use a lifecycle rule to delete non-current versions that are older than 30 days. If you delete a file, it does not go away - it becomes non-current until 30 days have gone by.

This is closer to what I was thinking too. Although note that with this rule, the 30 days are from the initial file creation, not from the time it was deleted. So any 30 day + old file will be immediately deleted.

**Comment by ddietterich** [comment]
Oh, I didn't realize that it was from create time. That isn't the desired behavior either. Hmmm...

**Comment by fboulnois** [comment]
@pshapiro4broad , @ddietterich . Ok, how about just turning on bucket versioning at minimum? Then we'll be able to recover older versions if necessary. We can add a lifecycle policy separately.

**Comment by ddietterich** [comment]
@pshapiro4broad I think you are mistaken. You can make a lifecycle rule based on DaysSinceNoncurrentTime which will retain an object for N days after it becomes non-current - so you can say "save it 30 days after it is deleted".

**Review by RoriCremer — APPROVED** [review]

File: src/main/java/bio/terra/service/resourcemanagement/google/GoogleBucketService.java [hunk 2]
```diff
@@ -276,6 +281,7 @@ public void updateBucketMetadata(String bucketName, String flightId) {
      * @return a reference to the bucket as a GCS Bucket object
      */
     private Bucket newCloudBucket(GoogleBucketResource bucketResource) {
+        boolean doVersioning = Arrays.stream(env.getActiveProfiles()).noneMatch(env -> env.contains("test"));
```
Comment (RoriCremer) [inline src/main/java/bio/terra/service/resourcemanagement/google/GoogleBucketService.java:284]: does this stream need to be collected?

Comment (fboulnois) [reply src/main/java/bio/terra/service/resourcemanagement/google/GoogleBucketService.java:284]: I think you only need to `collect` when you need to mutate and reduce the stream into something else.

File: src/main/java/bio/terra/service/resourcemanagement/google/GoogleBucketService.java [hunk 3]
```diff
@@ -285,6 +291,7 @@ private Bucket newCloudBucket(GoogleBucketResource bucketResource) {
             // See here for possible values: http://g.co/cloud/storage/docs/storage-classes
             .setStorageClass(StorageClass.REGIONAL)
             .setLocation(gcsConfiguration.getRegion())
+            .setVersioningEnabled(doVersioning)
```
Comment (nmalfroy) [inline src/main/java/bio/terra/service/resourcemanagement/google/GoogleBucketService.java:294]: I think it makes sense to turn on versioning but from what I can tell, this doesn't keep users from deleting objects.  Is that ok?

**Review by nmalfroy — APPROVED** [review]
talked offline.  lgtm!

**Comment by pshapiro4broad** [comment]
> @pshapiro4broad , @ddietterich . Ok, how about just turning on bucket versioning at minimum? Then we'll be able to recover older versions if necessary. We can add a lifecycle policy separately.

That is OK but with versioning enabled and no lifecycle policy in place, objects will never be deleted until the bucket is deleted. Also note that unlike retention, it doesn't protect against a bucket being deleted, such as during a 30 day retention period.

**Comment by pshapiro4broad** [comment]
@ddietterich I wonder if `DaysSinceNoncurrentTime` is new, I don't remember seeing it when I looked at this last, a few years ago.

**Pull request merged**